Add EMS Actuator for View Factor to Ground

bpark1327 opened the pull request:
Pull request overview
---------------------
 - Fixes #8476
 - Add EMS Actuator for View Factor to Ground
 - Addresses https://github.com/NREL/EnergyPlus/issues/8277#issuecomment-760529734


### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [x] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

bpark1327 committed:
add view factor to ground actuator.

bpark1327 committed:
Merge remote-tracking branch 'upstream/develop' into ems_actuator_shading

bpark1327 committed:
add a unit test

bpark1327 committed:
cleanup EMSMananger.unit.cc

mjwitte commented on doc/ems-application-guide/src/ems-actuators/thermal-envelope.tex:
```diff
@@ -55,6 +55,20 @@ \subsection{Surface Construction State}\label{surface-construction-state}
+
+\begin{itemize}
+\item
+  ``Outdoor Air Drybulb Temperature.''~ This is the dryblub temperature of the ambient outdoor air exposed to the surface, in degrees C. This is the temperature used for surface convection heat transfer boundary conditions on the outdoor, outside-face, of the surface exposed to exterior environment (unless it is raining).
```
Typo: dryblub

mjwitte commented on doc/ems-application-guide/src/ems-actuators/thermal-envelope.tex:
```diff
@@ -55,6 +55,20 @@ \subsection{Surface Construction State}\label{surface-construction-state}
+  ``Outdoor Air Wind Direction.''~ This is the wind direction of the outdoor air exposed to the surface, in degree. This is the direction used for some surface convection heat transfer correlations for the outside-face of surfaces exposed to wind.
+\item
+  ``View Factor To Ground.''~ This is the view factor to ground of the outdoor air exposed to the surface. 
+
```
Need `\end{itemize}` here.

mjwitte commented on doc/ems-application-guide/src/ems-actuators/thermal-envelope.tex:
```diff
@@ -55,6 +55,20 @@ \subsection{Surface Construction State}\label{surface-construction-state}
+\item
+  ``Outdoor Air Wind Direction.''~ This is the wind direction of the outdoor air exposed to the surface, in degree. This is the direction used for some surface convection heat transfer correlations for the outside-face of surfaces exposed to wind.
+\item
+  ``View Factor To Ground.''~ This is the view factor to ground of the outdoor air exposed to the surface. 
```
Would be good to mention here that this view factor is used for both solar and radiant exchange (if I remember correctly) and to mention the best calling point to use. `BeginZoneTimestepAfterInitHeatBalance` is that last calling point before `InitSurfaceHeatBalance` is called where the actuators are used.  `BeginZoneTimestepBeforeInitHeatBalance` could also be used.

bpark1327 replied:
I realized that I used `BeginTimestepBeforePredictor` in my test file. Does it have to be either `BeginZoneTimestepAfterInitHeatBalance` or `BeginZoneTimestepBeforeInitHeatBalance`?

mjwitte replied:
If I'm following everything correctly, with `BeginTimestepBeforePredictor` I think you'll find that the actual impact on the view factor will lag the EMS action by one zone timestep.

bpark1327 replied:
Aha, I see. I probably didn't see the impact because I only considered a seasonal shading coefficient variation. If I considered hourly variation (during the daytime), I would probably see some differences.

mjwitte reviewed (commented):
@bpark1327 
1. Code and changes look OK.
2. Doc changes are mostly OK - see comments below.
2. Please merge in develop for one last round of CI tests.
3. Do you have an idf file I can use to test the new actuator?
4. @Myoldmopar CI did not build docs here?
5. @Myoldmopar Unrelated to this PR, the err diffs here reveal some [existing API warnings](http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2021-01/d6cf92a2a7dc8ca3489d39792c4315f6f3bea957-2e34bd8ac6488886336710acecc6a6ebc9a1f7b2/PythonPluginUserDefined5ZoneAirCooled/x86_64-MacOS-10.15-clang-11.0.0/eplusout.err.diff.html) in the CI testfiles. Are those warnings expected?

bpark1327 committed:
update the documentation

bpark1327 committed:
Merge remote-tracking branch 'upstream/develop' into ems_actuator_shading

bpark1327 commented:
@mjwitte Thanks for your comments. I addressed your comments. I've attached the IDF you can use to test the new actuator. 
[ems-viewfactortoground-actuator.idf.txt](https://github.com/NREL/EnergyPlus/files/5906869/ems-viewfactortoground-actuator.idf.txt)
I changed the file extension of the IDF to .txt because Github does not allow me to attach an IDF file. Please change the file extension to IDF before you test it.  
Please let me know if you have any questions.

mjwitte commented on doc/ems-application-guide/src/ems-actuators/thermal-envelope.tex:
```diff
@@ -55,6 +55,21 @@ \subsection{Surface Construction State}\label{surface-construction-state}
+\item
+  ``Outdoor Air Wind Direction.''~ This is the wind direction of the outdoor air exposed to the surface, in degree. This is the direction used for some surface convection heat transfer correlations for the outside-face of surfaces exposed to wind.
+\item
+  ``View Factor To Ground.''~ This is the view factor to ground of the outdoor air exposed to the surface. This is the view factor used for solar and radiant heat exchange. This is useful from the calling point named BeginZoneTimestepAfterInitHeatBalance and BeginZoneTimestepBeforeInitHeatBalance.
```
One last doc change: "This is the view factor to ground of the outside face of he surface. This is the view factor used for ground diffuse solar and for radiant heat exchange."

mjwitte reviewed (approved):
@bpark1327 I've tested the new feature with a few variations and it seems to be working as expected. Please make the doc change noted below and merge in develop one more time

@mitchute or @Myoldmopar - does a new actuator require any corresponding changes in the PluginManager?

mitchute commented:
> does a new actuator require any corresponding changes in the PluginManager?

Not for new actuators. You will if you are adding new API functionality.

bpark1327 committed:
Another doc change

bpark1327 committed:
Merge remote-tracking branch 'upstream/develop' into ems_actuator_shading

mjwitte commented:
CI results look good except windows. Other pull requests show similar failures on windows, so not holding this up for that.
